### PR TITLE
Adding Persistence Data Support in Snorql

### PR DESCRIPTION
**Pull request by prajain12**
## Description

Snorq lwill  now support data persistence using Quartz Framework
1. User can configure wether a metric supports historical data or not.
2. User can set a minimum data store interval allowed.
3. User can perform CRUD operations on data recording jobs.
4. Reading historical data supports pagination

## Motivation and Context

A DBA needs to analyze historical data for database maintenance. Data Persistence enables DBA to configure triggers which are used to store data at particular intervals.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
## Backward compatibility

- [x] API
- [x] Client
- [x] Models
- [x] DB

### CODE REVIEW

File: snorql-framework/pom.xml [hunk 1]
```diff
@@ -63,6 +63,39 @@
...
+            <version>2.3.2</version>
+        </dependency>
+        <dependency>
+            <groupId>org.quartz-scheduler</groupId>
+            <artifactId>quartz-jobs</artifactId>
+            <version>2.3.2</version>
+        </dependency>
+        <dependency>
+            <groupId>com.microsoft.sqlserver</groupId>
+            <artifactId>mssql-jdbc</artifactId>
```
Comment (lokeshdevnani) [inline snorql-framework/pom.xml:78]: why is this required?

Comment (lokeshdevnani) [reply snorql-framework/pom.xml:78]: We should ideally not be mandating including any specific db driver in `snorql-framework`

File: snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/DataPersistenceJob.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     * Function executed when a data recording trigger is fired
+     * The function does the following:
+     * 1. Generates a random runID
+     * 2. Receives merged data map (contains data required for execution)
+     * 3. Fetches metric data
+     * 4. Calls the function to store data in historical data store
+     */
+    override fun execute(context: JobExecutionContext) {
+        try {
+            print("Quartz Job execution started!!")
```
Comment (lokeshdevnani) [inline snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/DataPersistenceJob.kt:28]: use logger

File: snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/JobManager.kt [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                    .withDescription(description)
+//                    .startAt(startFrom)
+                    .usingJobData(jobDataMap)
+                    .withSchedule(
+                        SimpleScheduleBuilder.simpleSchedule()
+                            .withIntervalInSeconds(intervalInSeconds)
+                            .repeatForever()
+                    )
+                    .endAt(endAt)
+                    .build()
```
Comment (lokeshdevnani) [inline snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/JobManager.kt:157]: This code seems to be repeating. Can this be extracted to a separate method?

Comment (prajain12) [reply snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/JobManager.kt]: Resolved.
Added  function buildTriggerObject in JobManager to generate a Simple Trigger Object

File: snorql-framework/pom.xml (hunk 1 above)
Comment (prajain12) [reply snorql-framework/pom.xml]: Post testing confirmed that Quartz does not require a JDBC driver explicitly for which this dependency was intended.
Removed mssql-jdbc dependency

File: snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/DataPersistenceJob.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            val dataRecorded = HistoricalDatabaseSchemaDTO(
+                runId = runID,
+                timestamp = Timestamp(System.currentTimeMillis()),
+                metricId = metricInput.metricId,
+                databaseName = metricInput.databaseName,
+                source = SnorqlConstants.DATA_PERSISTENCE_GROUP_NAME,
+                metricInput = SnorqlConstants.objectMapper.writeValueAsString(metricInput), // metricInput,
+                metricOutput = SnorqlConstants.objectMapper.writeValueAsString(metricOutput) // metricOutput
+            )
+            val storageId = SnorqlConstants.HISTORICAL_DATA_BUCKET_ID
```
Comment (lokeshdevnani) [inline snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/DataPersistenceJob.kt:48]: should this be part of the framework considering the storage data source is configurable?

**Review by lokeshdevnani — COMMENTED** [review]
Just a few doubts, rest looks fine

File: snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/DataPersistenceJob.kt (hunk 4 above)
Comment (prajain12) [reply snorql-framework/src/main/kotlin/com/udaan/snorql/framework/job/DataPersistenceJob.kt:48]: Added it to JobManager initialisation  flow.
By default HISTORICAL_DATA_BUCKET_ID is set to "sql_monitoring_historical_data" but it can be changed according to the user by passing HISTORICAL_DATA_BUCKET_ID in property file during initialisation

File: snorql-extensions/src/main/resources/sql-monitoring-conf.json [hunk 5]
```diff
@@ -4,6 +4,9 @@
...
     },
     "supportsHistorical": true,
+    "persistDataOptions": {
+      "minimumRepeatInterval": 2
```
Comment (lokeshdevnani) [inline snorql-extensions/src/main/resources/sql-monitoring-conf.json:8]:
```suggestion
      "minimumRepeatInterval": 10
```

File: snorql-extensions/src/main/resources/sql-monitoring-conf.json [hunk 6]
```diff
@@ -34,6 +43,9 @@
       "main": "dbcc show_statistics (:tableName, :indexName) with STAT_HEADER"
     },
     "supportsHistorical": true,
+    "persistDataOptions": {
+      "minimumRepeatInterval": 7200
```
Comment (lokeshdevnani) [inline snorql-extensions/src/main/resources/sql-monitoring-conf.json:47]:
```suggestion
      "minimumRepeatInterval": 3600
```

**Review by lokeshdevnani — APPROVED** [review]

**Review by lakshmi-udaan — APPROVED** [review]
